From ee103388e0dc3a29103aafa1cf8808d81ebf4a1f Mon Sep 17 00:00:00 2001
From: Sivasubramanaiyan Padmanaban <sivasubramaniyan.padmanaban@analog.com>
Date: Fri, 14 Aug 2026 11:27:41 +0530
Subject: [PATCH] media: i2c: nv_adsd3500: enable streaming before starting
 FSYNC PWM

- Send the STREAM-ON command before enabling the FSYNC PWM output.
- One frame-duration delay to allow the sensor to enter streaming mode.
- Ensure PWM-triggered synchronization starts only after streaming is active.

Signed-off-by: Sivasubramanaiyan Padmanaban <sivasubramaniyan.padmanaban@analog.com>
---
 drivers/media/i2c/nv_adsd3500.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/media/i2c/nv_adsd3500.c b/drivers/media/i2c/nv_adsd3500.c
index 8e9d67e1..4473cc20 100644
--- a/drivers/media/i2c/nv_adsd3500.c
+++ b/drivers/media/i2c/nv_adsd3500.c
@@ -1273,7 +1273,12 @@ static int adsd3500_start_streaming(struct adsd3500 *adsd3500)
 {
 	int ret;
 
+	ret = regmap_write(adsd3500->regmap, STREAM_ON_CMD, STREAM_ON_VAL);
+	if (ret < 0)
+		dev_err(adsd3500->dev, "Write of STREAM-ON command failed.\n");
+
 	if(adsd3500->pwm_fsync != NULL && adsd3500->curr_sync_mode == PWM_TRIGGER){
+		msleep(5);
 		ret = pwm_enable(adsd3500->pwm_fsync);
 		if (ret){
 			dev_err(adsd3500->dev, "Could not enable FSYNC PWM\n");
@@ -1281,10 +1286,6 @@ static int adsd3500_start_streaming(struct adsd3500 *adsd3500)
 		}
 	}
 
-	ret = regmap_write(adsd3500->regmap, STREAM_ON_CMD, STREAM_ON_VAL);
-	if (ret < 0)
-		dev_err(adsd3500->dev, "Write of STREAM-ON command failed.\n");
-
 	return ret;
 }
 
-- 
2.25.1

